sim/core/vengeance.go

```diff
@@ -0,0 +1,104 @@
+
+		// Decay existing bonus by half of the rate
+		decay := VengeanceAPDecayRate / 2
+		tracker.apBonus -= (decay * tracker.apBonus)
+
+		// Add 5% of damage taken in the last 2 seconds
+		tracker.apBonus += 0.05 * tracker.eligibleDamage
+
+		// 4.3.0 change: the vengeance AP buff is always at least 33% of the incoming
+		// damage if the tank has been hit in the last 2 seconds
+		baseAPBonus := tracker.eligibleDamage / 3.0
+		tracker.apBonus = math.Max(tracker.apBonus, baseAPBonus)
+	} else {
+		// No hits in the last 2 seconds - apply full decay
+		tracker.apBonus -= (VengeanceAPDecayRate * tracker.recentMaxAPBonus)
+	}
+
+	// TODO: verify that stam actually shows up in this formula, currently implemented as simcraft has it
+	apBonusMax := character.GetStat(stats.Stamina) + 0.1*character.GetStat(stats.Health)
+	tracker.apBonus = Clamp(tracker.apBonus, 0, apBonusMax)
+
+	tracker.recentMaxAPBonus = math.Max(tracker.apBonus, tracker.recentMaxAPBonus)
+
+	if sim.Log != nil {
+		character.Log(sim, "Updated Vengeance for %s: Eligible Damage(%f) | AP Bonus(%f)", character.Name, tracker.eligibleDamage, tracker.apBonus)
+	}
+
+	tracker.eligibleDamage = 0
+
+	// Update character stats
+	character.AddStatDynamic(sim, stats.AttackPower, -tracker.prevAPBonus)
+	character.AddStatDynamic(sim, stats.AttackPower, tracker.apBonus)
+}
+
+// To use: add a VengeanceTracker member to your spec-specific struct (e.g ProtWarrior, BloodDeathKnight, etc) then call this
+// with your class's specific Vengeance spell ID
+func ApplyVengeanceEffect(character *Character, tracker *VengeanceTracker, spellID int32) {
+
+	// For sanity
+	tracker.prevAPBonus = 0
+	tracker.apBonus = 0
+	tracker.eligibleDamage = 0
+	tracker.recentMaxAPBonus = 0
+
+	vengAura := MakePermanent(character.RegisterAura(Aura{
+		Label:    "Vengeance",
+		Duration: NeverExpires,
+		ActionID: ActionID{SpellID: spellID}, // Different specs use different spell IDs even though the effect is the same
+		OnSpellHitTaken: func(aura *Aura, sim *Simulation, spell *Spell, result *SpellResult) {
+			if result.Outcome.Matches(OutcomeVengeanceTriggers) {
+				// Vengeance is based on the taken damage amount after mitigation
+				// TODO: check how this treats dodge/parry/miss
+				// https://worldofwarcraft.blizzard.com/en-us/news/1293873/tanking-with-a-vengeance seems to suggest a string of dodges will let it fall off
+				// but simc's implementation retriggers vengeance on _any_ attack, even dodge/parry/miss.
+				// I can't find any patch notes or other resources that support one or the other though
+				tracker.lastAttackedTime = sim.CurrentTime
+				tracker.eligibleDamage += result.Damage
+			}
+		},
+	}))
+
+	// Vengeance "ticks" every 2 seconds to update the AP buff
+	character.RegisterResetEffect(func(sim *Simulation) {
+		StartPeriodicAction(sim, PeriodicActionOptions{
+			Period: time.Second * 2,
+			OnAction: func(sim *Simulation) {
+				UpdateVengeance(sim, character, tracker, vengAura)
+			},
+		})
+	})
```

Comment:
Make sure to test if this action is not stacking with every iteration as you are activating it on every reset. I dont remember if we automatically cancel all PAs at the end of a run (we should but leaving this comment as a sanity check)

Reply:
PAs do get cleared before each run here: https://github.com/wowsims/cata/blob/621230372254ac470d5df72537ce1d951ca80a7d/sim/core/sim.go#L377

This is also copying how a couple other similar effects were implemented in Wrath (warrior's Anger Management rage tick, priest's Rapture mana tick) so it seems it should still be okay.